src/service/ingestion/mod.rs

```diff
@@ -368,7 +368,18 @@ pub fn apply_stream_functions(
             value = apply_vrl_fn(runtime, vrl_runtime, &value, org_id, stream_name);
         }
     }
-    flatten::flatten_with_level(value, get_config().limit.ingest_flatten_level)
+
+    let flatten_level = get_config().limit.ingest_flatten_level;
+
+    if let Value::Array(v) = value {
```

Comment:
This should ideally be merged into `flatten::flatten_with_level` function. This would avoid repeating the same code in other places the same behavior is desired.